beeline.go

```diff
@@ -38,21 +39,39 @@ type Config struct {
 	// events. Default sampling is at the trace level - entire traces will be
 	// kept or dropped. default: 1 (meaning no sampling)
 	SampleRate uint
-	// SamplerHook is a function that will get run with the contents of each
-	// event just before sending the event to Honeycomb. Register a function
+	// SamplerHook is a function call that will be invoked with the contents of
+	// each event just before sending the event to Honeycomb. Register a function
 	// with this config option to have manual control over sampling within the
 	// beeline. The function should return true if the event should be kept and
 	// false if it should be dropped.  If it should be kept, the returned
 	// integer is the sample rate that has been applied. The SamplerHook
 	// overrides the default sampler. Runs before the PresendHook.
 	SamplerHook func(map[string]interface{}) (bool, int)
-	// PresendHook is a function call that will get run with the contents of
+	// PresendHook is a function call that will be invoked with the contents of
 	// each event just before sending them to Honeycomb. The function registered
 	// here may mutate the map passed in to add, change, or drop fields from the
 	// event before it gets sent to Honeycomb. Does not get invoked if the event
 	// is going to be dropped because of sampling. Runs after the SamplerHook.
 	PresendHook func(map[string]interface{})
-
+	// TraceHeaderParserHook is a function call that will be invoked with the context
+	// and HTTP request on all incoming HTTP requests. The function registered must
+	// return a list of HTTPPropagator objects that will be used to extract trace
+	// context from headers attached to the incoming request. Each propagator will
+	// be used, in sequence, and any conflicting information will be squashed by
+	// the later propagator. Decisions about which propagators to return can be
+	// made based on information in the context or HTTP request. The function may
+	// also return an empty array, which will result in no incoming headers being
+	// parsed.
+	TraceHTTPHeaderParserHook func(ctx context.Context, r *http.Request) []trace.HTTPPropagator
```

Comment:
Should add to this comment that if left undefined, Honeycomb trace headers will be honored.

Reply:
Should add that these functions are only invoked if you use one of Honeycomb's wrappers (either `hnynethttp` or one of the web framework wrappers). 

I think this bolsters the argument that this config might be best done at the level of the wrapper rather than at the `beeline` level.

Reply:
> This is the first introduction of anything specific to HTTP in to the beeline itself. It makes me wonder if these hooks should be part of the middleware wrappers that are automatic instrumentation specific to HTTP rather than in the main beeline package.
> So far the way those breakdown:
> 
> * beeline - dirt simple interface to adding tracing data to an application
> * trace - more complete and powerful and complicated interface to traces
> * autoinstrumentation - hooks for specific packages, including HTTP
> 
> In the interests of "dirt simple" it might be right to move HTTP into the beeline but I'd like to talk through it a bit more.

I'm going to write up some example code that uses some possible interface definitions - I think it'd be easier to discuss this if we had something tangible to point at.